core/common/src/Buffer.kt

```diff
@@ -624,20 +571,20 @@ public class Buffer : Source, Sink {
 
         val builder = StringBuilder(len * 2 + if (size > maxPrintableBytes) 1 else 0)
 
-        var curr = head!!
-        var bytesWritten = 0
-        var pos = curr.pos
-        while (bytesWritten < len) {
-            if (pos == curr.limit) {
-                curr = curr.next!!
-                pos = curr.pos
-            }
-
-            val b = curr.data[pos++].toInt()
-            bytesWritten++
-
-            builder.append(HEX_DIGIT_CHARS[(b shr 4) and 0xf])
-                .append(HEX_DIGIT_CHARS[b and 0xf])
+        UnsafeBufferOperations.iterate(this) { ctx, head ->
+            var bytesWritten = 0
+            var seg: Segment? = head
+            do {
+                seg!!
+                var idx = 0
+                while (bytesWritten < len && idx < seg.size) {
+                    val b = ctx.getUnchecked(seg, idx++)
+                    bytesWritten++
+                    builder.append(HEX_DIGIT_CHARS[(b shr 4) and 0xf])
```

Comment:
Same here, append byte-by-byte

Reply:
`Buffer.toString` is not expected to be used for anything but debugging (as it can only capture a prefix of a buffer)